### PR TITLE
Show static members of matching types in completion

### LINKED ISSUE #28603: Intellisense should suggest static members of target type
(Visual Studio 15.7.5)

Intellisense should suggest static members of the target type which return an instance of the target type.

After typing `Foo(e`, wanting to pass an empty immutable array:

image

After typing `Foo(e` (ReSharper):

image

Visual Studio, after typing `Foo(e)`:

image

With ReSharper, after typing `Foo(e)`:

image

This is also quite useful for hand-rolled discriminated unions. (`foo = Foo.Bar`, `foo = Foo.Baz(3)`, etc.)

### PR DESCRIPTION
**Pull request by sharwell**
image

image

image

image

Closes #28603

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 1]
```diff
@@ -114,14 +115,23 @@ private static async Task HandleSingleTypeAsync(CompletionContext context, Seman
                 type = typeArg;
             }
 
+            var showTypeOnlyIfMembers = false;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:118]: this name wasn't really clear to me.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 2]
```diff
@@ -114,14 +115,23 @@ private static async Task HandleSingleTypeAsync(CompletionContext context, Seman
...
+            INamedTypeSymbol? lazyEnclosingNamedType = null;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:119]: this was super confusing to me.  i couldn't see how this value was being written before being read.  i'm not a fan of doing tihs and then having a local function setting this.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:119]: i don't see value in this being lazy.  it's not expensive to compute, and tons of other services will have already asked for this.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 3]
```diff
@@ -174,6 +182,84 @@ private static async Task HandleSingleTypeAsync(CompletionContext context, Seman
...
+                        AddTypeIfNotAdded();
+                        var memberDisplayName = $"{displayText}.{property.Name}";
+                        context.AddItem(SymbolCompletionItem.CreateWithSymbolId(
+                            displayText: memberDisplayName,
+                            displayTextSuffix: "",
+                            symbols: ImmutableArray.Create<ISymbol>(property),
+                            rules: CompletionItemRules.Default,
+                            contextPosition: position,
+                            sortText: memberDisplayName,
+                            filterText: memberDisplayName));
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:232]: logic seems almost identical.  seems like we literally just use the type/nameof the symbol.  can we share this?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 4]
```diff
@@ -174,6 +182,84 @@ private static async Task HandleSingleTypeAsync(CompletionContext context, Seman
...
+                    }
+                    else
+                    {
+                        continue;
+                    }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:237]: i'd just remove this.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 5]
```diff
@@ -248,5 +334,13 @@ protected override Task<CompletionDescription> GetDescriptionWorkerAsync(Documen
                 ? completionListType
                 : null;
         }
+
+        private static INamedTypeSymbol? TryGetTypeWithStaticMembers(ITypeSymbol type)
+        {
+            if (type.TypeKind == TypeKind.Struct || type.TypeKind == TypeKind.Class)
+                return type as INamedTypeSymbol;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:341]: does this match records?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 1 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:118]: When I woke up this morning I had a new thought to remove this.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 5 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:341]: There is no `TypeKind.Record` so I'm assuming yes?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 4 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:237]: ⏱️ Cleaned up as part of other work in progress

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 3 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:232]: ⏱️ Cleaned up as part of other work in progress

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 2 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:119]: ⏱️ Making eagerly computed

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 5 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:315]: ➡️ Test now added

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs [hunk 6]
```diff
@@ -29,15 +29,26 @@ static void Main(string[] args)
...
 enum Colors
 {
     Red,
     Blue,
     Green,
 }
 ";
-            await VerifyItemExistsAsync(markup, "Colors");
+            var colorsLike = @"
+readonly struct Colors
```
Comment (heaths) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs:42]: Given the implementation covers both fields and properties, might consider adding a type defining static get-only properties.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 7]
```diff
@@ -174,6 +193,47 @@ private static async Task HandleSingleTypeAsync(CompletionContext context, Seman
...
+                        staticSymbol = property;
+                        symbolType = property.Type;
+                    }
+                    else
+                    {
+                        // Only fields and properties are supported for static member matching
+                        continue;
+                    }
+
+                    if (!SymbolEqualityComparer.Default.Equals(type, symbolType)
```
Comment (heaths) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:219]: Should this also be limited to read-only fields and properties (get-only)? The original intent seems to be to treat such members as (immutable) enums, though off hand I don't see any major reason to also limit this to read-write static fields or properties if used as, say, defaults. Just wanted to raise the question, though.

**Review by heaths — COMMENTED** [review]
Thank you! This will be great for the many ["extensible enums"](https://azure.github.io/azure-sdk/dotnet_implementation.html#dotnet-enums) we have in the Azure SDKs!

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 8]
```diff
@@ -1854,8 +1854,8 @@ public class C
                   <Document><![CDATA[
 public class C
 {
-    public C(int Alice, int Bob) { }
-    public C(string ignored) { }
+    public C(object Alice, object Bob) { }
+    public C(object ignored) { }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:1858]: so this case is def interesting.  why did it break?  can you clarify?  i have a guess, but i don't want to go down a wrong path of assumptions.

**Review by CyrusNajmabadi — CHANGES_REQUESTED** [review]
need clarification on potential behavioral change and if the new behavior is desirable.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs (hunk 6 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs:42]: This should be covered by `TestProperties()`:

https://github.com/dotnet/roslyn/blob/3475981a50c03c5bb060c4abcaa3518fdf6fbcfb/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProviderTests.cs#L1916-L1938

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 8 above)
Comment (sharwell) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:1858]: Enum types always ignored the option under test here. The test needed to be updated because `int` (`int.MaxValue`) and `string` (`string.Empty`) both act like enums for triggering here.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:1858]: so my pref would be to take that test and doc the new behavior, and then add a new test for the old scenario personally.  not a huge deal though.

Comment (CyrusNajmabadi) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb:1858]: but that does seem like an acceptable chance here.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 7 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:219]: Right now it doesn't restrict to read-only, but that shouldn't be a problem.

File: src/Features/Core/Portable/Completion/CompletionRules.cs [hunk 9]
```diff
@@ -160,7 +160,7 @@ public CompletionRules WithSnippetsRule(SnippetsRule snippetsRule)
         private static readonly ImmutableArray<char> s_defaultCommitKeys = ImmutableArray.Create(
                 ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
                 ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
-                '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
+                '~', '=', '<', '>', '?', '#', '\'', '\"', '\\');
```
Comment (genlu) [inline src/Features/Core/Portable/Completion/CompletionRules.cs:163]: Why this change?

Comment (sharwell) [reply src/Features/Core/Portable/Completion/CompletionRules.cs:163]: The easiest way to see the bug was this scenario (even before my change). Start with this (caret at `$$`):

```
string @this;
string str;
str = $$
```

Then press <kbd>Ctrl</kbd>+<kbd>Space</kbd> followed by <kbd>@</kbd>. Completion should allow the `@` to be typed as the start of an identifier, but right now it cannot.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs [hunk 10]
```diff
@@ -174,6 +193,47 @@ private static async Task HandleSingleTypeAsync(CompletionContext context, Seman
                         filterText: memberDisplayName));
                 }
             }
+            else if (enclosingNamedType is not null)
```
Comment (genlu) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:196]: Does it work with top level statement?

**Comment by sharwell** [comment]
Experimentation revealed a whole bunch of cases where it's a problem to automatically show the completion list. For the new cases, will modify this to only show the completion list after a character is typed.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs (hunk 10 above)
Comment (sharwell) [reply src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs:205]: ➡️ Yes, at least for the case of typing this:

```
int.Parse(empty)
```

File: src/Features/Core/Portable/Completion/CompletionRules.cs (hunk 9 above)
Comment (sharwell) [reply src/Features/Core/Portable/Completion/CompletionRules.cs]: ➡️ This change was reverted.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb (hunk 8 above)
Comment (sharwell) [reply src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: ➡️ This change was reverted.

**Comment by heaths** [comment]
> For the new cases, will modify this to only show the completion list after a character is typed.

Would this work as designed if someone were to force the completion list using, for example (bindings vary), Ctrl+Space? I.e. the completion list would show, but would it initially show/have selected one of these well-known possible values?

**Comment by sharwell** [comment]
Yes, explicit <kbd>Ctrl</kbd>+<kbd>Space</kbd> will bring up completion with these items included.

**Comment by heaths** [comment]
> included

Included, or selected by default. E.g. in the case of enums, the enum is selected automatically after the completion list is filled (seems to be a *very short* delay):

image

Just trying to understand the expectation here. The hope was that these work like enums, though I can imagine some of the issues you might've hit above, and appreciate having to workaround that to maintain development productivity and consistency.

**Comment by sharwell** [comment]
@heaths They will be included, but do not influence the selection algorithm. Depending on other factors, an element from this collection may or may not be selected by default.
